README.md
# Autoprefixer

<img align="right" width="94" src="http://ai.github.io/autoprefixer/logo.svg" title="Autoprefixer logo by Anton Lovchikov">

Parse CSS and add vendor prefixes to CSS rules using values
from the [Can I Use](http://caniuse.com/).

Write your CSS rules without vendor prefixes (in fact, forget about them
entirely):

```css
:fullscreen a {
    transition: transform 1s
}
```

Process your CSS by Autoprefixer:

```js
var prefixed = autoprefixer.process(css).css;
```

It will use the data on current browser popularity and properties support
to apply prefixes for you:

```css
:-webkit-full-screen a {
    -webkit-transition: -webkit-transform 1s;
    transition: transform 1s
}
:-moz-full-screen a {
    transition: transform 1s
}
:-ms-fullscreen a {
    transition: transform 1s
}
:fullscreen a {
    -webkit-transition: -webkit-transform 1s;
    transition: transform 1s
}
```

Twitter account for news and releases:
[@autoprefixer](https://twitter.com/autoprefixer).

Sponsored by [Evil Martians](http://evilmartians.com/).
Based on [PostCSS](https://github.com/ai/postcss) framework.

## Features

### Forget about prefixes

The best tool is a tool you can't see and one that does the work for you.
This is the main idea behind Autoprefixer.

Autoprefixer interface is simple: just forget about vendor prefixes
and write normal CSS according to latest W3C specs. You don’t need
a special language (like Sass) or special mixins.

Because Autoprefixer is a postprocessor for CSS,
you can also use it with preprocessors, such as Sass, Stylus or LESS.

### Actual data from Can I Use

Autoprefixer uses the most recent data from [Can I Use](http://caniuse.com/),
understands which browsers are actual and popular and adds only the necessary
vendor prefixes.

It also cleans your CSS from old prefixes (like prefixed `border-radius`,
produced by many CSS libraries):

```css
a {
  -webkit-border-radius: 5px;
  border-radius: 5px
}
```

compiles to:

```css
a {
  border-radius: 5px
}
```

### Rewrite syntax

Flexbox or gradients have different syntaxes in different browsers
(sometimes you need to recalculate angles, sometimes you need 2 old properties
instead of new one), but Autoprefixer hides this from you.

Just code by latest W3C specs and Autoprefixer will produce the code
for old browsers:

```css
a {
  display: flex;
}
```

compiles to:

```css
a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -moz-box;
  display: -ms-flexbox;
  display: flex
}
```

### Fast

Autoprefixer is about 16 times faster than Compass and 8 times faster
than Stylus.

On a Core i7 with 10 GB of RAM and SSD, benchmark with GitHub styles is:

```
~/Dev/autoprefixer$ ./node_modules/.bin/cake bench
Load GitHub styles
Autoprefixer: 522 ms
Compass:      4901 ms (9.4 times slower)
Stylus:       4544 ms (8.7 times slower)
```

Unlike [-prefix-free](http://leaverou.github.io/prefixfree/), Autoprefixer
compiles CSS once on deploy and doesn’t hit client-side performance.

## Browsers

You can specify the browsers you want to target in your project:

```js
autoprefixer("last 1 version", "> 1%", "explorer 7").process(css).css;
```

* `last 2 versions` is last versions for each browser. Like “last 2 versions”
  [strategy](http://support.google.com/a/bin/answer.py?answer=33864) in
  Google.
* `last 2 chrome versions` is last versions of specify browser.
* `> 5%` is browser versions, selected by global usage statistics.
* `firefox > 20` and `firefox >= 20` is Firefox versions newer, that 20.
* `none` don’t set any browsers to clean CSS from any vendor prefixes.
* You can also set browsers directly.

Blackberry and stock Android browsers will not be used in `last n versions`.
You can add them by name:

```js
autoprefixer("last 1 version", "blackberry 10", "android 4").process(css).css;
```

Browsers names (case insensitive):
* `android` for old Android stock browser.
* `blackberry` or `bb` for Blackberry browser.
* `chrome` for Google Chrome.
* `firefox` or `ff` for Mozilla Firefox.
* `explorer` or `ie` for Internet Explorer.
* `ios` for iOS Safari.
* `opera` for Opera.
* `safari` for desktop Safari.

By default, Autoprefixer uses `> 1%, last 2 versions, firefox 24, opera 12.1`:
* Firefox 24 is a latest [ESR].
* Opera 12.1 will be in list until Opera supports non-Blink 12.x branch.

[ESR]: http://www.mozilla.org/en/firefox/organizations/faq/

## Source Map

Autoprefixer will generate source map, if you set `map` option to `true`.

You must set input and output CSS files paths (by `from` and `to` options)
to generate correct map.

```js
var result = autoprefixer.process(css, {
    map:  true,
    from: 'main.css',
    to:   'main.out.css'
});

result.css //=> Prefixed CSS
result.map //=> Source map content

fs.writeFileSync('main.out.css.map', result.map);
```

Autoprefixer can also modify previous source map (for example, from Sass
compilation). Just set original source map content (as string or JS object)
to `map` option:

```js
var result = autoprefixer.process(css, {
    map:   fs.readFileSync('main.sass.css.map'),
    from: 'main.sass.css',
    to:   'main.min.css'
});

result.map //=> Source map from main.sass to main.min.css
```

## Debug

You can check which browsers are selected and which properties will be prefixed:

```js
info = autoprefixer("last 1 version").info();
console.log(info);
```

Or by CLI command:

```sh
autoprefixer -i
```

## FAQ

### Does it add polyfills for old browsers?

No. Autoprefixer only adds prefixes, not polyfills. There are two reasons:

1. Prefixes and polyfills are very different and need a different API.
   Two separate libraries would be much better.
2. Most of IE polyfills are very bad for client perfomance. They use slow hacks
   and old IEs is mostly used on old hardware. Most of CSS 3 features that is
   only used for styling should be ignored in old IEs as it is recommended in
   Graceful Degradation.

### Why Autoprefixer doesn’t add prefixes to `border-radius`?

Developers are often surprised by how few prefixes are required today.
If Autoprefixer doesn’t add prefixes to your CSS, check if they’re still
required on [Can I Use](http://caniuse.com/).

If a prefix is required, but Autoprefixer doesn’t add it or adds it
incorrectly, please
[report an issue](https://github.com/ai/autoprefixer/issues/new)
and include your source CSS and expected output.

### Why Autoprefixer doesn’t support `appearance`?

Instead of `transition`, `appearance` property is not a part of
any specification. So there is no `appearance`, only `-moz-appearance`
and `-webkit-appearance`. Quote from [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/-moz-appearance):

> Do not use this property on Web sites: not only is it non-standard, but its
> behavior changes from one browser to another. Even the keyword `none` does not
> have the same behavior on each form element across different browsers, and
> some do not support it at all.

## Usage

### Grunt

You can use the
[grunt-autoprefixer](https://github.com/nDmitry/grunt-autoprefixer)
plugin for Grunt. Install the npm package and add it to Gruntfile:

```js
grunt.loadNpmTasks('grunt-autoprefixer');
```

### Compass

If you use Compass binary to compile your styles, you can easily integrate
Autoprefixer with it. Install `autoprefixer-rails` gem:

```
gem install autoprefixer-rails
```

and add post-compile hook to `config.rb`:

```ruby
require 'autoprefixer-rails'
require 'csso'

on_stylesheet_saved do |file|
  css = File.read(file)
  File.open(file, 'w') do |io|
    io << AutoprefixerRails.compile(css)
  end
end
```

You can set browsers array as second argument in `AutoprefixerRails.compile`.

### Stylus

If you use Stylus CLI, you can add Autoprefixer by
[autoprefixer-stylus](https://github.com/jenius/autoprefixer-stylus) plugin.

Just install npm package and use it in `-u` option:

```
stylus -u autoprefixer-stylus file.css
```

### Ruby on Rails

Add [autoprefixer-rails](https://github.com/ai/autoprefixer-rails) gem
to `Gemfile` and write CSS in a usual way:

```ruby
gem "autoprefixer-rails"
```

### Ruby

You can integrate Autoprefixer into your Sprockets environment
by `autoprefixer-rails` gem:

```ruby
AutoprefixerRails.install(sprockets_env)
```

or process CSS from plain Ruby:

```ruby
prefixed = AutoprefixerRails.compile(css)
```

### Prepros

If you want to build your assets in GUI, try
[Prepros](http://alphapixels.com/prepros/). Just set “Auto Prefix CSS”
[checkbox](https://f.cloud.github.com/assets/3478693/930798/faa29892-0016-11e3-8901-87850de7aed2.jpg)
in right panel.

<img src="http://alphapixels.com/prepros/static/img/prepros.jpg" width="550" height="340" />

### Mincer

To use Autoprefixer in [Mincer](https://github.com/nodeca/mincer),
install `autoprefixer` npm package and enable it:

```js
environment.enable('autoprefixer');
```

### Middleman

Add [middleman-autoprefixer](https://github.com/porada/middleman-autoprefixer)
gem to `Gemfile`:

```ruby
gem "middleman-autoprefixer"
```

and activate the extension in your project’s `config.rb`:

```ruby
activate :autoprefixer
```

### Node.js

Use `autoprefixer` npm package:

```js
var autoprefixer = require('autoprefixer');
var prefixed     = autoprefixer.process(css).css;
```

### PHP

You can use Autoprefixer in PHP by
[autoprefixer-php](https://github.com/vladkens/autoprefixer-php) library:


```php
$autoprefixer = new Autoprefixer();
$css          = 'a { transition: transform 1s }';
$prefixed     = $autoprefixer->compile($css);
```

### JavaScript

You can use Autoprefixer in the browser or a non-Node.js runtime
with [standalone version](https://raw.github.com/ai/autoprefixer-rails/master/vendor/autoprefixer.js).

### PostCSS

Autoprefixer can be also used as a [PostCSS](https://github.com/ai/postcss)
processor, so you can combine it with other processors and parse CSS only once:

```js
postcss().
    use( autoprefixer(['> 1%', 'opera 12.5']).postcss ).
    use( compressor ).
    process(css);
```

### Sublime Text

You can process your styles directly in Sublime Text with the
[sublime-autoprefixer](https://github.com/sindresorhus/sublime-autoprefixer)
plugin.


### Others

You can use the `autoprefixer` binary to process CSS files using
any assets manager:

```
sudo npm install --global autoprefixer
autoprefixer *.css
```

See `autoprefixer -h` for help.

## In-package Update

It’s highly recommended that you always use the latest version of Autoprefixer.
If by any chance you or your company are not able to update the package
(e.g. in case of long test periods before any library updates), you can still
update the very browser data that Autoprefixer fetches from Can I Use:

```
autoprefixer --update
```

Note that the in-package update doesn’t get any code fixes nor the
implementation of new features. It just keeps the browser popularity and
support data up to date, and adds new browser versions.


ChangeLog
== 0.8 (Unbowed, Unbent, Unbroken)
* Add more browsers to defaults ("> 1%, last 2 versions, ff 17, opera 12.1"
  instead of just "last 2 browsers").
* Keep vendor prefixes without unprefixed version (like vendor-specific hacks).
* Convert gradients to old WebKit syntax (actual for Android 2.3).
* Better support for several syntaxes with one prefix (like Flexbox and
  gradients in WebKit).
* Add intrinsic and extrinsic sizing values support.
* Remove never existed prefixes from common mistakes (like -ms-transition).
* Add Opera 17 data.
* Fix selector prefixes order.
* Fix browser versions order in inspect.

20130903:
* Fix old WebKit gradients convertor on rgba() colors.
* Allow to write old direction syntax in gradients.

20130906:
* Fix direction syntax in radial gradients.
* Don’t prefix IE filter with modern syntax.

20130911:
* Fix parsing property name with spaces.

20130919:
* Fix processing custom framework prefixes (by Johannes J. Schmidt).
* Concat outputs if several files compiled to one output.
* Decrease standalone build size by removing unnecessary Binary class.
* iOS 7 is moved to released versions.
* Clean up binary code (by Simon Lydell).

20130923:
* Firefox 24 is moved to released versions.

20131001:
* Add support for grab, grabbing, zoom-in and zoom-out cursor values.

20131006:
* Chrome 30 is moved to released versions.

20131007:
* Don’t add another prefixes in rule with prefixed selector.

20131009:
* Opera 17 is moved to released versions.

20131015:
* Fix converting multiple gradients to old webkit syntax (by Aleksei Androsov).

20131017:
* Fix @host at-rule parsing.

20131020:
* IE 11 and Andrid 4.3 is moved to released versions.
* Add Opera 18 data.
* Add @namespace support.
* Sort browser versions in data file.

20131029:
* Add Safari 6.1 data.
* Add fx alias for Firefox.

20131104:
* Update Android future version to 4.4.
* Google Chrome 32 added to future versions list.
* Firefox 25 now is actual version, 27 and 28 added to future versions.
* Browsers statistics are updated.

20131205:
* Google Chrome 33 added to future releases list.
* Google Chrome 31 moved to current releases list.

20131209:
* Use old webkit gradients for old iOS and Safari (by Chad von Nau).
* Fix direction conversion for old webkit gradients (by Chad von Nau).
* Update browsers popularity statistics.

20131213:
* Firefox ESR in default browsers was changed to 24 version.
* Firefox 26 was moved to current releases list.
* Firefox 28 was added to future releases list.

== 0.7 (We Do Not Sow)
* Add vendor prefixes to selectors.
* Add ::selection and ::placeholder selectors support.
* Allow to load support data from Can I Use pull requests.
* Remove deprecated API.

20130806:
* Add hyphens support.

20130807:
* Add tab-size support.
* Add :fullscreen support.

20130808:
* Allow to select browser versions by > and >= operator.
* Fix flex properties in transition.

20130810:
* Add Firefox 25 data.

20130824:
* Add Chrome 31 and 30 data.
* Fix CSS comments parsing (by vladkens).

== 0.6 (As High As Honor)
* New faster API, which cache preprocessed data. Old API is deprecated.
* A lot of perfomance improvements.
* Add Opera 15 -webkit- prefix support.
* Update Chrome 29 and Safari 7 prefixes data.
* Add minor browsers in popularity select.
* Better syntax error messages.

20130721:
* Add Chrome 30 data.

20130728:
* Don’t remove non-standard -webkit-background-clip: text.
* Don’t remove IE hack on CSS parse.

20130729:
* Add Opera 16 data.
* Fix “Invalid range in character class” error on Firefox.

20130730:
* Fix correct clone comments inside keyframes (by Alexey Plutalov).
* Fix angle recalculation in gradients (by Roman Komarov).

20130731:
* Add border-image support.

== 0.5 (Ours is the Fury)
* Rewrite Autoprefixer to be more flexible.
* Use css, instead of Rework, to fix CSS parsing errors faster.
* Fix a lot of CSS parsing errors.

20130616:
* More useful message for CSS parsing errors.
* Remove old WebKit gradient syntax.
* Fix parsing error on comment with braces.

20130617:
* Remove old Mozilla border-radius.
* Don’t prefix old IE filter.
* Remove old background-clip, background-size and background-origin prefixes.
* Speed up regexps in values.
* Allow to hack property declarations.

20130625:
* Convert flexbox properties to 2009 and 2012 specifications.
* Improve messages on syntax errors.

20130626:
* Add Firefox 24 data.
* Add prefixes for font-feature-settings.

20130629:
* Fix convert flex properties to old box-flex.

== 0.4 (Winter Is Coming)
* Remove outdated prefixes.
* Add border-radius and box-shadow properties to database.
* Change degrees in webkit gradients.

20130515:
* Add old syntax in gradient direction.
* Add old syntax for display: flex.
* Update browser global usage statistics.

20130521:
* Add Firefox 23 data.

20130524:
* Add Chrome 29 data.

20130528:
* Fix compatibilty with Rework from git master.
* Add minor browsers to data, which can be selected only directly.

20130530:
* Add Opera 15 and iOS 6.1 data.
* Fix iOS versions in properties and values data.

20130603:
* Use latest Rework 0.15 with a lot of CSS parsing fixes.
* Update browsers usage statistics.

== 0.3 (Growing Strong)
* Rename `autoprefixer.filter()` to `autoprefixer.rework()`.
* Use own filters instead of Rework’s `prefix` and `prefixValue`.
* Smarter value prefixer without false match “order” in “border”.
* 40% faster.
* Don’t add unnecessary properties instead of Rework’s `prefixValue`.
* Don’t change properties order.
* Sort properties and values in inspect output.
* Add main to component config (by Jonathan Ong).
* Fix documentation (by Sergey Leschina and Mark Vasilkov).

20130424:
* Fix value override in prefixer.

20130427:
* Prefix several same values in one property.
* Fix Windows support in binary.
* Improve print errors in binary.

20130502:
* Don’t add -webkit- prefix to IE filter.
* Don’t duplicate prefixes on second run.

== 0.2 (Hear Me Roar!)
* Update parse libraries.

data/values.js
// Don't edit this files, because it's autogenerated.
// See updaters/ dir for generator. Run bin/update to update.

module.exports = {
    calc: {
        props: [
            "*"
        ],
        browsers: [
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "safari 6",
            "ios 6",
            "ios 6.1",
            "bb 10"
        ]
    },
    flex: {
        props: [
            "display"
        ],
        browsers: [
            "ie 10",
            "ff 2",
            "ff 3",
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "ff 16",
            "ff 17",
            "ff 18",
            "ff 19",
            "ff 20",
            "ff 21",
            "ff 3.5",
            "ff 3.6",
            "chrome 4",
            "chrome 5",
            "chrome 6",
            "chrome 7",
            "chrome 8",
            "chrome 9",
            "chrome 10",
            "chrome 11",
            "chrome 12",
            "chrome 13",
            "chrome 14",
            "chrome 15",
            "chrome 16",
            "chrome 17",
            "chrome 18",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "chrome 26",
            "chrome 27",
            "chrome 28",
            "chrome 29",
            "safari 4",
            "safari 5",
            "safari 6",
            "safari 7",
            "safari 3.1",
            "safari 3.2",
            "safari 5.1",
            "opera 15",
            "ios 3.2",
            "ios 4",
            "ios 4.1",
            "ios 4.2",
            "ios 4.3",
            "ios 5",
            "ios 5.1",
            "ios 6",
            "ios 6.1",
            "ios 7",
            "android 3",
            "android 4",
            "android 2.1",
            "android 2.2",
            "android 2.3",
            "android 4.1",
            "android 4.2",
            "bb 7",
            "bb 10"
        ],
        replace: function (string, prefix, rules) {
            if (prefix === '-webkit-') {
                rules.add('display', '-webkit-box');
                return '-webkit-' + string;
            } else if (prefix === '-moz-') {
                return prefix + 'box';
            } else if (prefix === '-ms-') {
                return prefix + 'flexbox';
            }
        }
    },
    "inline-flex": {
        props: [
            "display"
        ],
        browsers: [
            "ie 10",
            "ff 2",
            "ff 3",
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "ff 16",
            "ff 17",
            "ff 18",
            "ff 19",
            "ff 20",
            "ff 21",
            "ff 3.5",
            "ff 3.6",
            "chrome 4",
            "chrome 5",
            "chrome 6",
            "chrome 7",
            "chrome 8",
            "chrome 9",
            "chrome 10",
            "chrome 11",
            "chrome 12",
            "chrome 13",
            "chrome 14",
            "chrome 15",
            "chrome 16",
            "chrome 17",
            "chrome 18",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "chrome 26",
            "chrome 27",
            "chrome 28",
            "chrome 29",
            "safari 4",
            "safari 5",
            "safari 6",
            "safari 7",
            "safari 3.1",
            "safari 3.2",
            "safari 5.1",
            "opera 15",
            "ios 3.2",
            "ios 4",
            "ios 4.1",
            "ios 4.2",
            "ios 4.3",
            "ios 5",
            "ios 5.1",
            "ios 6",
            "ios 6.1",
            "ios 7",
            "android 3",
            "android 4",
            "android 2.1",
            "android 2.2",
            "android 2.3",
            "android 4.1",
            "android 4.2",
            "bb 7",
            "bb 10"
        ]
    },
    "linear-gradient": {
        props: [
            "background",
            "background-image"
        ],
        browsers: [
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "ff 3.6",
            "chrome 4",
            "chrome 5",
            "chrome 6",
            "chrome 7",
            "chrome 8",
            "chrome 9",
            "chrome 10",
            "chrome 11",
            "chrome 12",
            "chrome 13",
            "chrome 14",
            "chrome 15",
            "chrome 16",
            "chrome 17",
            "chrome 18",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "safari 4",
            "safari 5",
            "safari 6",
            "safari 7",
            "safari 5.1",
            "opera 12",
            "opera 11.1",
            "opera 11.5",
            "opera 11.6",
            "ios 3.2",
            "ios 4",
            "ios 4.1",
            "ios 4.2",
            "ios 4.3",
            "ios 5",
            "ios 5.1",
            "ios 6",
            "ios 6.1",
            "ios 7",
            "android 3",
            "android 4",
            "android 2.1",
            "android 2.2",
            "android 2.3",
            "android 4.1",
            "android 4.2",
            "bb 7",
            "bb 10"
        ],
        replace: function (string, prefix) {
            var regexp;
            regexp = /to\s+(top|bottom)?\s*(left|right)?/ig;
            string = string.replace(regexp, function (_, ver, hor) {
                    var direct;
                    direct = [];
                    if (ver === 'top') {
                        direct.push('bottom');
                    }
                    if (ver === 'bottom') {
                        direct.push('top');
                    }
                    if (hor === 'right') {
                        direct.push('left');
                    }
                    if (hor === 'left') {
                        direct.push('right');
                    }
                    return direct.join(' ');
                });
            regexp = /(repeating-)?(linear|radial)-gradient/gi;
            string = string.replace(regexp, prefix + '$&');
            if (prefix !== '-webkit-') {
                return string;
            }
            regexp = /((repeating-)?(linear|radial)-gradient\()\s*(-?\d+deg)?/ig;
            return string.replace(regexp, function (_0, gradient, _1, _2, deg) {
                    if (deg) {
                        deg = parseInt(deg);
                        deg += 90;
                        if (deg > 360) {
                            deg -= 360;
                        }
                        return gradient + deg + 'deg';
                    } else {
                        return gradient;
                    }
                });
        }
    },
    "radial-gradient": {
        props: [
            "background",
            "background-image"
        ],
        browsers: [
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "ff 3.6",
            "chrome 4",
            "chrome 5",
            "chrome 6",
            "chrome 7",
            "chrome 8",
            "chrome 9",
            "chrome 10",
            "chrome 11",
            "chrome 12",
            "chrome 13",
            "chrome 14",
            "chrome 15",
            "chrome 16",
            "chrome 17",
            "chrome 18",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "safari 4",
            "safari 5",
            "safari 6",
            "safari 7",
            "safari 5.1",
            "opera 12",
            "opera 11.1",
            "opera 11.5",
            "opera 11.6",
            "ios 3.2",
            "ios 4",
            "ios 4.1",
            "ios 4.2",
            "ios 4.3",
            "ios 5",
            "ios 5.1",
            "ios 6",
            "ios 6.1",
            "ios 7",
            "android 3",
            "android 4",
            "android 2.1",
            "android 2.2",
            "android 2.3",
            "android 4.1",
            "android 4.2",
            "bb 7",
            "bb 10"
        ],
        replace: function (string, prefix) {
            var regexp;
            regexp = /to\s+(top|bottom)?\s*(left|right)?/ig;
            string = string.replace(regexp, function (_, ver, hor) {
                    var direct;
                    direct = [];
                    if (ver === 'top') {
                        direct.push('bottom');
                    }
                    if (ver === 'bottom') {
                        direct.push('top');
                    }
                    if (hor === 'right') {
                        direct.push('left');
                    }
                    if (hor === 'left') {
                        direct.push('right');
                    }
                    return direct.join(' ');
                });
            regexp = /(repeating-)?(linear|radial)-gradient/gi;
            string = string.replace(regexp, prefix + '$&');
            if (prefix !== '-webkit-') {
                return string;
            }
            regexp = /((repeating-)?(linear|radial)-gradient\()\s*(-?\d+deg)?/ig;
            return string.replace(regexp, function (_0, gradient, _1, _2, deg) {
                    if (deg) {
                        deg = parseInt(deg);
                        deg += 90;
                        if (deg > 360) {
                            deg -= 360;
                        }
                        return gradient + deg + 'deg';
                    } else {
                        return gradient;
                    }
                });
        }
    },
    "repeating-linear-gradient": {
        props: [
            "background",
            "background-image"
        ],
        browsers: [
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "ff 3.6",
            "chrome 4",
            "chrome 5",
            "chrome 6",
            "chrome 7",
            "chrome 8",
            "chrome 9",
            "chrome 10",
            "chrome 11",
            "chrome 12",
            "chrome 13",
            "chrome 14",
            "chrome 15",
            "chrome 16",
            "chrome 17",
            "chrome 18",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "safari 4",
            "safari 5",
            "safari 6",
            "safari 7",
            "safari 5.1",
            "opera 12",
            "opera 11.1",
            "opera 11.5",
            "opera 11.6",
            "ios 3.2",
            "ios 4",
            "ios 4.1",
            "ios 4.2",
            "ios 4.3",
            "ios 5",
            "ios 5.1",
            "ios 6",
            "ios 6.1",
            "ios 7",
            "android 3",
            "android 4",
            "android 2.1",
            "android 2.2",
            "android 2.3",
            "android 4.1",
            "android 4.2",
            "bb 7",
            "bb 10"
        ],
        replace: function (string, prefix) {
            var regexp;
            regexp = /to\s+(top|bottom)?\s*(left|right)?/ig;
            string = string.replace(regexp, function (_, ver, hor) {
                    var direct;
                    direct = [];
                    if (ver === 'top') {
                        direct.push('bottom');
                    }
                    if (ver === 'bottom') {
                        direct.push('top');
                    }
                    if (hor === 'right') {
                        direct.push('left');
                    }
                    if (hor === 'left') {
                        direct.push('right');
                    }
                    return direct.join(' ');
                });
            regexp = /(repeating-)?(linear|radial)-gradient/gi;
            string = string.replace(regexp, prefix + '$&');
            if (prefix !== '-webkit-') {
                return string;
            }
            regexp = /((repeating-)?(linear|radial)-gradient\()\s*(-?\d+deg)?/ig;
            return string.replace(regexp, function (_0, gradient, _1, _2, deg) {
                    if (deg) {
                        deg = parseInt(deg);
                        deg += 90;
                        if (deg > 360) {
                            deg -= 360;
                        }
                        return gradient + deg + 'deg';
                    } else {
                        return gradient;
                    }
                });
        }
    },
    "repeating-radial-gradient": {
        props: [
            "background",
            "background-image"
        ],
        browsers: [
            "ff 4",
            "ff 5",
            "ff 6",
            "ff 7",
            "ff 8",
            "ff 9",
            "ff 10",
            "ff 11",
            "ff 12",
            "ff 13",
            "ff 14",
            "ff 15",
            "ff 3.6",
            "chrome 4",
            "chrome 5",
            "chrome 6",
            "chrome 7",
            "chrome 8",
            "chrome 9",
            "chrome 10",
            "chrome 11",
            "chrome 12",
            "chrome 13",
            "chrome 14",
            "chrome 15",
            "chrome 16",
            "chrome 17",
            "chrome 18",
            "chrome 19",
            "chrome 20",
            "chrome 21",
            "chrome 22",
            "chrome 23",
            "chrome 24",
            "chrome 25",
            "safari 4",
            "safari 5",
            "safari 6",
            "safari 7",
            "safari 5.1",
            "opera 12",
            "opera 11.1",
            "opera 11.5",
            "opera 11.6",
            "ios 3.2",
            "ios 4",
            "ios 4.1",
            "ios 4.2",
            "ios 4.3",
            "ios 5",
            "ios 5.1",
            "ios 6",
            "ios 6.1",
            "ios 7",
            "android 3",
            "android 4",
            "android 2.1",
            "android 2.2",
            "android 2.3",
            "android 4.1",
            "android 4.2",
            "bb 7",
            "bb 10"
        ],
        replace: function (string, prefix) {
            var regexp;
            regexp = /to\s+(top|bottom)?\s*(left|right)?/ig;
            string = string.replace(regexp, function (_, ver, hor) {
                    var direct;
                    direct = [];
                    if (ver === 'top') {
                        direct.push('bottom');
                    }
                    if (ver === 'bottom') {
                        direct.push('top');
                    }
                    if (hor === 'right') {
                        direct.push('left');
                    }
                    if (hor === 'left') {
                        direct.push('right');
                    }
                    return direct.join(' ');
                });
            regexp = /(repeating-)?(linear|radial)-gradient/gi;
            string = string.replace(regexp, prefix + '$&');
            if (prefix !== '-webkit-') {
                return string;
            }
            regexp = /((repeating-)?(linear|radial)-gradient\()\s*(-?\d+deg)?/ig;
            return string.replace(regexp, function (_0, gradient, _1, _2, deg) {
                    if (deg) {
                        deg = parseInt(deg);
                        deg += 90;
                        if (deg > 360) {
                            deg -= 360;
                        }
                        return gradient + deg + 'deg';
                    } else {
                        return gradient;
                    }
                });
        }
    }
};
